pkg/strcase/strcase.go

```diff
@@ -0,0 +1,70 @@
+// Package strcase implements functions to convert a camelCase UTF-8 string into various cases.
+//
+// New delimiters will be inserted based on the following transitions:
+//   - On any change from lowercase to uppercase letter.
+//   - On any change from number to uppercase letter.
+package strcase
+
+import (
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// Delimited converts a string to delimited.lower.case, here using `.` as delimiter.
+func Delimited(s string, d rune) string {
+	return convert(s, unicode.LowerCase, d)
+}
+
+// ScreamingDelimited converts a string to DELIMITED.UPPER.CASE, here using `.` as delimiter.
+func ScreamingDelimited(s string, d rune) string {
+	return convert(s, unicode.UpperCase, d)
+}
+
+// Snake converts a string to snake_case.
+func Snake(s string) string {
+	return Delimited(s, '_')
+}
+
+// ScreamingSnake converts a string to SCREAMING_SNAKE_CASE.
+func ScreamingSnake(s string) string {
+	return ScreamingDelimited(s, '_')
+}
+
+// convert converts a camelCase UTF-8 string into various cases.
+// _case must be unicode.LowerCase or unicode.UpperCase.
+func convert(s string, _case int, d rune) string {
+	if len(s) == 0 {
+		return s
+	}
+
+	n := strings.Builder{}
+	n.Grow(len(s) + 2) // Allow adding at least 2 delimiters without another allocation.
+
+	var pos int
+	var prevRune int32
+
+	for _, r := range s[pos:] {
+		n.WriteRune(unicode.To(_case, r))
+
+		pos += utf8.RuneLen(r)
+		prevRune = r
+
+		// This loop just extracts the first rune of the string,
+		// without needing to allocate a new slice of runes as
+		// range iterates over the runes of the string.
+		break
+	}
+
+	for _, r := range s[pos:] {
+		if unicode.IsUpper(r) && (unicode.IsNumber(prevRune) || unicode.IsLower(prevRune)) {
+			n.WriteRune(d)
+		}
+
+		n.WriteRune(unicode.To(_case, r))
+
+		prevRune = r
+	}
```

Comment:
This can be simplified into a single loop (and there's `rune` as an alias for `int32` which is better fitting here):

```suggestion
	var prevRune rune

	for i, r := range s {
		if i > 0 && unicode.IsUpper(r) && (unicode.IsNumber(prevRune) || unicode.IsLower(prevRune)) {
			n.WriteRune(d)
		}

		n.WriteRune(unicode.To(_case, r))

		prevRune = r
	}
```

The `i > 0` check wouldn't even be necessary as `(unicode.IsNumber(0) || unicode.IsLower(0)) == false`, but I don't think it hurts to make this a bit more explicit.

Bonus: this fixes a panic with input `"\xff"`:
```
panic: runtime error: slice bounds out of range [3:1]
```

If the input starts with invalid UTF-8, `for ... range` will iterate over it as `r = '�'` which is 3 bytes long, so `pos += utf8.RuneLen(r)` will add 3 which may be beyond the end of the string.

This requires the current `{"\xff\xfe\xfd", "�"}` test case to `{"\xff\xfe\xfd", "���"}` instead. This looks fine to me, at least I don't see a reason why these bytes should be treated as a single character. Treating it as three characters also matches the Go iteration behavior (well, that's basically the reason why it becomes three unicode replacement characters).

Reply:
Makes sense.